### ISSUE TITLE
msgpack TimeoutErrors

### ISSUE THREAD

**Issue by amadeus** [Shougo/ddc.vim#39]
# Problems summary
After a while in a session, I start seeing msgpack timeout errors, and was wondering if there was an expectation here.  I am pretty sure it's due to various sources never resolving their completion results.  Is this something source authors need to ensure they handle properly? Or is this an error that ddc should suppress?

## Expected
Not to see TimeoutErrors in the message logs

## Environment Information

 * ddc version (SHA1): `e392213`

 * OS:
 macOS

 * neovim/Vim `:version` output:
 macVim, 8.2, 1-3376 

## Provide a minimal init.vim/vimrc with less than 50 lines (Required!)
I am continuing to dig in and see if I can find a specific source of the problem, but figured I would start the dialogue about whether it's a plugin or ddc issue.

## Screenshot (if possible)
<img width="1393" alt="CleanShot 2021-08-30 at 12 39 08" src="https://user-images.githubusercontent.com/83376/131395374-9c3b28e2-6b6b-423d-87f1-999775ded878.png">

**Comment by Shougo** [comment]
It seems deno process communication problem than ddc problem ...
But I need the minimal vimrc to reproduce it for me.
We have not ESP skills.

The minimal vimrc and the reproduce instruction is required.
If you don't upload them, I will close the issue unfortunately.

**Comment by Shougo** [comment]
https://deno.land/x/msgpack_rpc@v3.1.0/response_waiter.ts

The default response timeout = `10000` ms = `10` s.
I think it is long.  Why it does not respond within the time?

**Comment by Shougo** [comment]
It seems ddc-ale problem from your screenshot.  I think the source does not respond.

**Comment by amadeus** [comment]
Sorry, I did not supply a minimal vimrc earlier, I only had a hunch of the problem. Anyways, I think now I've got a pretty clear picture of what is going on, after testing a bunch of stuff today throughout the day.

The minimal vimrc would be:

```
call plug#begin('~/.vim/bundle')
Plug 'dense-analysis/ale'
Plug 'vim-denops/denops.vim'
Plug 'Shougo/ddc.vim'
Plug 'statiolake/ddc-ale'
call plug#end()
call ddc#custom#patch_global('sources', ['ale'])
call ddc#enable()
```

However, it's a bit unclear about what the correct fix would be. A summary of the issue is - Ale does not always respond with results, and in cases of rapid typing input, ddc pushes many consecutive requests to ale, but ale only remembers and fires the callback on the most recent one.

The code path starts here:
https://github.com/statiolake/ddc-ale/blob/master/denops/ddc/sources/ale.ts#L35-L41
which calls into here:
https://github.com/statiolake/ddc-ale/blob/master/autoload/ddc/ale.vim#L8-L17
Which in Ale, reaches into here:
https://github.com/dense-analysis/ale/blob/d098124e59ba38b0fcd97b41eec4a6de81bc09ab/autoload/ale/completion.vim#L833
At which point, it overrides a few buffer local variables that hold onto the callback and the request info, BUT only for the last call into GetCompletions (i.e. each subsequent call to GetCompletions overwrites all callback and info data for the previous call):
https://github.com/dense-analysis/ale/blob/d098124e59ba38b0fcd97b41eec4a6de81bc09ab/autoload/ale/completion.vim#L843-L845
https://github.com/dense-analysis/ale/blob/d098124e59ba38b0fcd97b41eec4a6de81bc09ab/autoload/ale/completion.vim#L867-L875

So basically we end up with requests into Ale's completion system, but we ultimately only respond with the latest callback, which leaves a bunch listeners that ultimately time out.

It would probably be a much bigger change to `fix` Ale to hold onto all requests and have all callbacks fire, but it would also create a lot of additional work for DDC to process requests that might be out of date. I did hack a bit on the ddc-ale plugin to vastly minimize request spam, but it's not ideal since if the number of results are huge, the max number might kick in and hurt your chances of finding what you are looking for:
https://gist.github.com/amadeus/9a438bc725d5892d7eedb754008c9928

From an overall ddc architecture discussion though, it does kinda feel like ddc should be built in a way that the requests from these services may never come back, and at least give plugin devs a way to handle or catch a timeout, instead of just logging to echom.

**Comment by Shougo** [comment]
I will implement that ddc catches `TimeoutError`.

**Comment by Shougo** [comment]
I think `ddc-ale` catch the error or `SetTimeout()` though...

**Comment by amadeus** [comment]
> I think `ddc-ale` can catch the error or `SetTimeout()` though...

The plugin used to have a setTimeout in place to resolve earlier, but that can cause weird issues if it ends up resolving while you are autocompleting something else, then the results end up getting wiped out or changed

**Comment by amadeus** [comment]
I wonder if DDC should have a way to know the results that have come in are delayed, and no longer relevant, and can just throw them away

**Comment by Shougo** [comment]
I have skipped the timeout.  And I have added the check for it.

**Comment by amadeus** [comment]
Unfortunately it looks like there are still some of these TimeoutErrors triggering through another system. The minimal vimrc posted above I think is still applicable:

<img width="1393" alt="CleanShot 2021-08-31 at 14 40 58" src="https://user-images.githubusercontent.com/83376/131579549-43013f02-3a8e-48c4-a9c2-6cf788e28fcf.png">

However, I think it might have to do with another promise timing out somewhere, not applicable to the fetching stuff we've been discussing here?

**Comment by Shougo** [comment]
Really?????  I have added the exception handler.
Why it does not work?

Please investigate why it is occurred.

And I think ddc-ale should use callback like `ddc-nvim-lsp` instead.
The current implementation seems blocked.

**Comment by Shougo** [comment]
I think ddc cannot work for it.  No magic.

**Comment by amadeus** [comment]
Still digging, will report if I can find anything

**Comment by Shougo** [comment]
`ddc-vim-lsp` had the same problem and fixed by source implementation fix.

https://github.com/shun/ddc-vim-lsp

**Comment by amadeus** [comment]
Is the idea that you coalesce similar fetches to re-use the previous promise? I have a suspicion that this issues it not related to fetches anymore, since your fix looks good. I think it might be coming from some other await level code...

**Comment by Shougo** [comment]
> Is the idea that you coalesce similar fetches to re-use the previous promise? 

Please explain it more clearly.

**Comment by Shougo** [comment]
I think to catch `TimeoutErrors` is useless.  I should remove the catch.
It should be fixed in source implementation side.

**Comment by Shougo** [comment]
Please check the catch is called or not.

https://github.com/Shougo/ddc.vim/blob/main/denops/ddc/app.ts#L174

**Comment by amadeus** [comment]
> I should remove the catch.

I still think it's better user experience to catch timeout error and not emit them (unless perhaps the user enables a debug mode or something). The inherent nature of these async systems is that they may never return, and punishing the user for them seems not great? Turning on debug mode or setting a global variable however, can be very useful to have them emit to catch problems.

As an aside, I am still trying to track down the source of these timeout errors, I do not believe they are coming from gatherResults, but from some other `await` usage.  They don't seem to happen right away like before, so I think your fixes are still good for the initial but I found here.  Right now, I am just wrapping a variety of different await cases and adding some catches with logs to see if I can find the offending ones (but it seems to take a long time to get the errors, and I haven't figured out a direct way to repro, so it's going slow).

It would be nice if these errors had a bit more descriptive a trace that would let us know where they are emitting from, is there perhaps something on the denops side I could do to make this easier to debug?

**Comment by amadeus** [comment]
I still don't know the cause yet, but I've uncovered a bit more info (not sure if it's helpful or not).

I modified the code here:
https://github.com/vim-denops/denops.vim/blob/main/denops/@denops-private/service.ts#L116

To include more date when it throws, like so:
```ts
  async dispatch(name: string, fn: string, args: unknown[]): Promise<unknown> {
    try {
      const plugin = this.#plugins.get(name);
      if (!plugin) {
        throw new Error(`No plugin '${name}' is registered`);
      }
      return await plugin.session.call(fn, ...args);
    } catch (e) {
      // NOTE:
      // Vim/Neovim does not handle JavaScript Error instance thus use string instead
      const logger: string[] = [];
      args.forEach((item) => {
        if (typeof item === "string") {
          logger.push(item);
        } else if (typeof item === 'number') {
          logger.push(`${item}`);
        } else if (typeof item === "object" || Array.isArray(item)) {
          logger.push(JSON.stringify(item));
        } else  {
          logger.push('unknown item')
        }
      });
      throw `${name}:${fn} -- ${logger.join(':')} --${e.stack ?? e.toString()}`;
    }
  }
```

And the timeout errors I am getting are:
```
[denops] ddc:onEvent -- TextChangedI --TimeoutError: the process didn't complete in time
[denops]     at https://deno.land/x/msgpack_rpc@v3.1.0/response_waiter.ts:49:25
[denops]     at fire (deno:ext/timers/01_timers.js:477:7)
[denops]     at handleTimerMacrotask (deno:ext/timers/01_timers.js:338:7)
```

**Comment by Shougo** [comment]
OK.  So please add the catch routine in `onEvent` and check where the error is occurred.

**Comment by amadeus** [comment]
Sorry for the delays, busy day, and I needed a good amount of time to test whether my solutions were working.

I think I managed to fix things in 99% of the cases now, and managed to fix them on the `ddc-ale` side of things via this PR here:

https://github.com/statiolake/ddc-ale/pull/4

TL;DR: `ddc-ale` was registering an async callback using `denops.call` - https://github.com/statiolake/ddc-ale/blob/master/denops/ddc/sources/ale.ts#L27-L31

Essentially I was seeing timeout errors related to the `once` handlers never getting fired. Anyways, I think my PR should fix the issues (there might still be a stray timeout here or there, not entirely positive I've solved it all yet, but this is the nature with highly async programming 😅 )

**Comment by Shougo** [comment]
I have improved `TimeoutError` catches.
ddc catches all of sources/filters `TimeoutError`.

Closing.

**Comment by amadeus** [comment]
🙏 yeet! thanks for hearing me out on all this, and thanks for adding those additional timeout safety checks! Muuuuuch appreciated!

Btw, really liking this plugin (despite a few of the growing pains so far, but it's beta after all, to be expected), and really happy you are open to feedback, and excited to see where it leads!

(Given what I am learning, I may have to dig in and make a file autocomplete source since you mentioned it was not something you wanted to take on 🤣)

**Comment by Shougo** [comment]
> Btw, really liking this plugin (despite a few of the growing pains so far, but it's beta after all, to be expected), and really happy you are open to feedback, and excited to see where it leads!

You can contribute it!

> (Given what I am learning, I may have to dig in and make a file autocomplete source since you mentioned it was not something you wanted to take on rofl)

Yeah, I don't want to implement complex file/path source.
You can create it instead.  I need to focus on plugin core development.

**Comment by Shougo** [comment]
file/path source is not implemented yet by users.
https://github.com/topics/ddc-vim

It means file/path source is very complicated source than others.
